### PR TITLE
Add FXIOS-10594 - Add attachments support to RS and pull `tracking-protection-lists-ios`

### PR DESCRIPTION
**Pull request by issammani**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-10594)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/23201)

## :bulb: Description
This PR:
- Adds fetching attachments to `Update_Remote_Settings.py`.
- Adds `tracking-protection-lists-ios` collection to config.
- Adds two fields to the config `fetch_attachments` and `save_records`.
- Runs the github action manually and adds the attachments for `tracking-protection-lists-ios`.
## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

**Comment by issammani** [comment]
cc @mattreaganmozilla ( and @nbhasin2 ) for the pp session today. I prepared some stuff 😄

**Comment by mattreaganmozilla** [comment]
@issammani This looks great so far, thanks for taking this on 👏

File: firefox-ios/Client/Assets/RemoteSettingsData/RemoteSettingsFetchConfig.json [hunk 1]
```diff
@@ -1,11 +1,19 @@
 {
-  "rules": [
+  "collections": [
```
Comment (nbhasin2) [inline firefox-ios/Client/Assets/RemoteSettingsData/RemoteSettingsFetchConfig.json:2]: don't forget to update [RemoteSettingsFetchConfig.swift](https://github.com/mozilla-mobile/firefox-ios/blob/main/firefox-ios/Client/Application/RemoteSettings/RemoteSettingsFetchConfig.swift) as you changed rules to collections

Comment (issammani) [reply firefox-ios/Client/Assets/RemoteSettingsData/RemoteSettingsFetchConfig.json:2]: Aha Thanks. I did in fact forget that 🙈

Comment (issammani) [reply firefox-ios/Client/Assets/RemoteSettingsData/RemoteSettingsFetchConfig.json:2]: Done in [7670e5f](https://github.com/mozilla-mobile/firefox-ios/pull/23199/commits/7670e5f9aabd407362a4f2457bad21f540f01981)

**Comment by lmarceau** [comment]
Just a quick question, I am out of context but we now commit the json disconnect files, like disconnect-block-advertising.json? Just want to understand what changed, as I remember we shouldn't commit, or at least make sure the disconnect files are not part of the app binary when we ship it.

File: .github/workflows/firefox-ios-update_remote_settings_data_script.yml [hunk 2]
```diff
@@ -44,7 +44,7 @@ jobs:
       - name: Add changes
         run: |
           git diff
-          git diff --quiet || git add firefox-ios/Client/Assets/RemoteSettingsData/RemotePasswordRules.json
+          git diff --quiet || git add firefox-ios/Client/Assets/RemoteSettingsData/
```
Comment (nbhasin2) [inline .github/workflows/firefox-ios-update_remote_settings_data_script.yml:47]: nice

File: firefox-ios/Client/Assets/RemoteSettingsData/Update_Remote_Settings.py [hunk 3]
```diff
@@ -28,34 +45,83 @@ def update_settings_file(tmp_file_path, target_file_path, name):
...
             print(f"Updated {target_file_path} with new rules for {name}.")
             return True
     else:
-        os.replace(tmp_file_path, target_file_path)
+        replace_file_with_dirs(tmp_file_path, target_file_path)
         print(f"Created new rules file {target_file_path} for {name}.")
         return True
 
+
+def fetch_records_attachments(records, collection, base_url):
```
Comment (nbhasin2) [inline firefox-ios/Client/Assets/RemoteSettingsData/Update_Remote_Settings.py:57]: nice!

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/RemoteSettings/RemoteSettingsFetchConfigTests.swift [hunk 4]
```diff
@@ -18,17 +18,17 @@ class RemoteSettingsFetchConfigTests: XCTestCase {
             return
         }
 
-        XCTAssertGreaterThan(config.rules.count, 0, "Expected more than 0 rules in the config")
+        XCTAssertGreaterThan(config.collections.count, 0, "Expected more than 0 collections in the config")
```
Comment (nbhasin2) [inline firefox-ios/firefox-ios-tests/Tests/ClientTests/RemoteSettings/RemoteSettingsFetchConfigTests.swift:21]:
```suggestion
        XCTAssertGreaterThan(
            config.collections.count,
            0,
            "Expected more than 0 collections in the config"
        )
```
⚠️ nit

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/RemoteSettings/RemoteSettingsFetchConfigTests.swift [hunk 5]
```diff
@@ -18,17 +18,17 @@ class RemoteSettingsFetchConfigTests: XCTestCase {
...
-            file: "./firefox-ios/Client/Assets/RemoteSettingsData/RemotePasswordRules.json",
+            url: "https://firefox.settings.services.mozilla.com/v1",
+            file: Optional("./firefox-ios/Client/Assets/RemoteSettingsData/RemotePasswordRules.json"),
             bucketID: "main",
-            collectionsID: "password-rules"
+            collectionID: "password-rules"
         )
 
-        XCTAssertTrue(config.rules.contains { $0 == expectedRule }, "Expected rule not found in the loaded config")
+        XCTAssertTrue(config.collections.contains { $0 == expectedCollection }, "Expected collection not found in the loaded config")
```
Comment (nbhasin2) [inline firefox-ios/firefox-ios-tests/Tests/ClientTests/RemoteSettings/RemoteSettingsFetchConfigTests.swift:31]:
```suggestion
        XCTAssertTrue(
            config.collections.contains { $0 == expectedCollection },
            "Expected collection not found in the loaded config"
        )
```
⚠️ nit

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/RemoteSettings/RemoteSettingsFetchConfigTests.swift [hunk 6]
```diff
@@ -37,12 +37,17 @@ class RemoteSettingsFetchConfigTests: XCTestCase {
...
+            XCTAssertFalse(collection.name.isEmpty, "Collection name should not be empty")
+            XCTAssertFalse(collection.url.isEmpty, "Collection URL should not be empty")
+            if collection.saveRecords ?? true {
+                XCTAssertFalse(collection.file!.isEmpty, "Collection file path should exist and not be empty")
+            } else {
+                XCTAssertNil(collection.file, "Collection file path should not exist")
+            }
+            
+            XCTAssertFalse(collection.bucketID.isEmpty, "Bucket ID should not be empty")
+            XCTAssertFalse(collection.collectionID.isEmpty, "Collection ID should not be empty")
```
Comment (nbhasin2) [inline firefox-ios/firefox-ios-tests/Tests/ClientTests/RemoteSettings/RemoteSettingsFetchConfigTests.swift:50]:
```suggestion
        for collection in config.collections {
            XCTAssertFalse(collection.name.isEmpty, "Collection name should not be empty for \(collection)")
            XCTAssertFalse(collection.url.isEmpty, "Collection URL should not be empty for \(collection)")
            
            if let file = collection.file, collection.saveRecords ?? true {
                XCTAssertFalse(file.isEmpty, "Collection file path should exist and not be empty for \(collection)")
            } else {
                XCTAssertNil(collection.file, "Collection file path should not exist for \(collection)")
            }

            XCTAssertFalse(collection.bucketID.isEmpty, "Bucket ID should not be empty for \(collection)")
            XCTAssertFalse(collection.collectionID.isEmpty, "Collection ID should not be empty for \(collection)")

            // Validate URL format
            XCTAssertNotNil(URL(string: collection.url), "Collection URL is invalid: \(collection.url)")
        }
```
⚠️ Swiftlint ! issue

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/RemoteSettings/RemoteSettingsFetchConfigTests.swift (hunk 4 above)
Comment (issammani) [reply firefox-ios/firefox-ios-tests/Tests/ClientTests/RemoteSettings/RemoteSettingsFetchConfigTests.swift:21]: Hmmm how did that slip but the test passed locally 🤔

Comment (issammani) [reply firefox-ios/firefox-ios-tests/Tests/ClientTests/RemoteSettings/RemoteSettingsFetchConfigTests.swift:21]: Aha linting issue haha I thought it was about changing `config.rules.count` 😂

**Comment by nbhasin2** [comment]
I gave you R+ but are you sure below is right?
image

**Comment by issammani** [comment]
> I gave you R+ but are you sure below is right?

Nice catch @nbhasin2. I updated the code in https://github.com/mozilla-mobile/firefox-ios/pull/23199/commits/6e4c6d63fe003836722ec0898c723065659e855f to save the attachments as is from the server. 

### Before https://github.com/mozilla-mobile/firefox-ios/pull/23199/commits/6e4c6d63fe003836722ec0898c723065659e855f

Screenshot 2024-11-20 at 20 07 37


### After https://github.com/mozilla-mobile/firefox-ios/pull/23199/commits/6e4c6d63fe003836722ec0898c723065659e855f

Screenshot 2024-11-20 at 20 08 23

**Review by nbhasin2 — APPROVED** [review]

**Review by lmarceau — APPROVED** [review]

**Comment by issammani** [comment]
Hey @isabelrios, @dragosb01 is OOO. Can you take a look please ? 😄

**Review by nbhasin2 — APPROVED** [review]

**Review by clarmso — APPROVED** [review]

**Pull request merged**